core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -486,56 +488,57 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
 
     Authority auth = Authority.fromString(uri.getAuthority());
     if (auth instanceof UnknownAuthority) {
-      // TODO(zac): In Alluxio 2.0 this warning will be upgraded to an exception
-      LOG.warn("Authority \"{}\" is unknown. The client will not be configured with this"
-          + " authority. The authority connection details will be loaded from your client"
-          + " configuration.",
-          auth);
-      mAlluxioHeader = getScheme() + ":///";
+      throw new IOException(String.format("Authority \"%s\" is unknown. The client can not be "
+              + "configured with this authority.", auth.toString()));
     }
 
     mUri = URI.create(mAlluxioHeader);
 
     Map<String, Object> uriConfProperties = getConfigurationFromUri(uri);
+    AlluxioProperties alluxioProps = ConfigurationUtils.defaults();
+    Subject subject = getHadoopSubject();
+    if (subject != null) {
+      LOG.debug("Using Hadoop subject: {}", subject);
+    } else {
+      LOG.debug("No Hadoop subject. Using FileSystem Context without subject.");
+    }
 
-    synchronized (INIT_LOCK) {
-      if (sInitialized) {
-        if (!connectDetailsMatch(uriConfProperties, conf)) {
-          LOG.warn(ExceptionMessage.DIFFERENT_CONNECTION_DETAILS.getMessage(
-              FileSystemContext.get().getMasterInquireClient().getConnectDetails()));
-          initializeInternal(uriConfProperties, conf);
-        }
-      } else {
-        initializeInternal(uriConfProperties, conf);
-      }
-      // Must happen inside the lock so that the global filesystem context isn't changed by a
+    synchronized (this) {
```

Comment:
What is `synchronized` protecting? initializing the state? If so, should config creation be protected or not?

Also, shouldn't `mInitialized` be checked while holding the lock?

Reply:
You're right. It should be checked. The method should never be called concurrently but because it's a public method it's possible that the user calls it twice. Adding the synchronized here prevents the state from getting mangled. However, on closer inspection is might be a better idea to expand the synchronization block up to the top of the method so that it's never even possible to initialize two `FileSystemContexts`.

i.e.

```
initialize() {
  synchronized (this) {
    if (!mInitialized) {
      // Do initialization (method body)
    } else {
       //throw exception 
    }
  }

}
```

Reply:
@gpang @ZacBlanco why do we need to support calling `initialize` twice? Is the expected behavior even well-defined?

Reply:
the behavior isn't defined but the point is that we should be able to handle it by either throwing an exception or re-initializing in the case that the method ever does get called twice. We have no guarantee that the user won't be able to call this method so we should clearly define how we handle the case where it _is_ called a second time (or concurrently)

Reply:
having well-defined behavior of our own sgtm, as long as we keep it as simple as possible (e.g. Preconditions check on volatile boolean)